### PR TITLE
raftstore: update lease before applying

### PR DESCRIPTION
**Pull request by BusyJay**
When changing to async applying, it's not easy to update lease from another thread. So we need to move lease updating from applying.

@siddontang @hhkbp2 PTAL

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -686,30 +712,104 @@ impl Peer {
...
+                // It's not leader anymore, we are safe to clear proposals. If it becomes leader
+                // again, the lease should be updated when election is finished, old proposals
+                // have no effect. Although this may allow old uuids to be reused before old
+                // callbacks are called.
+                self.proposals.clear();
+            }
+            for entry in committed_entries.iter().rev() {
+                // raft meta is very small, can be ignored.
+                self.raft_log_size_hint += entry.get_data().len() as u64;
+                to_be_updated = to_be_updated &&
```
Comment (hhkbp2) [inline src/raftstore/store/peer.rs]: It's fine to update leader lease only for the last entry with latest propose time.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -740,41 +840,54 @@ impl Peer {
...
-            self.pending_cmds.set_conf_change(cmd);
+            is_conf_change = true;
         } else if let Err(e) = self.propose_normal(req, metrics) {
             cmd_resp::bind_error(&mut err_resp, e);
-            cmd.call(err_resp);
+            cb(err_resp);
             return false;
+        }
+
+
```
Comment (hhkbp2) [inline src/raftstore/store/peer.rs]: Empty line.

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -740,41 +840,54 @@ impl Peer {
...
+        // Try to renew leader lease on every consistent read/write request.
+        meta.renew_lease_time = Some(clocktime::raw_now());
+
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: remove empty line

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -686,30 +712,104 @@ impl Peer {
...
 
         self.raft_group.advance(ready);
     }
 
+    /// Try to update lease.
+    ///
+    /// If the it can make sure that its lease is the latest lease, returns true.
+    fn maybe_update_lease(&mut self, term: u64, data: &[u8]) -> bool {
+        let mut req = RaftCmdRequest::new();
+        let propose_time = match req.merge_from_bytes(data)
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: can we avoid to decode the data twice, one here and the other in apply?

can we use term + index to find the propose time?

File: src/raftstore/store/peer.rs (hunk 1 above)
Comment (siddontang) [inline src/raftstore/store/peer.rs]: +1

**Comment by siddontang** [comment]
I think maybe we can use term + index to identify a command uniquely, no need `uuid`. (Now, the uuid is generated in storage in TiKV directly, seem unnecessary)

E.g, leader proposes a command to raft log at term 10 with index 10 and commits the log successfully, the position can't be changed. 
If the leader is down and another leader proposes a new command at index 10, the term will >= 11. 

/cc @BusyJay @hhkbp2

File: src/raftstore/store/peer.rs (hunk 1 above)
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: This is exactly what it's doing. If it update lease successfully, `maybe_update_lease` will not be called again.

Comment (siddontang) [inline src/raftstore/store/peer.rs]: seem you should use `||`

File: src/raftstore/store/peer.rs (hunk 4 above)
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Normally only the last entry will be decoded twice. The overhead can be omitted. Use term + index can find the propose time, but I still need to store uuid of every proposal already, so here just use uuid.

Comment (siddontang) [inline src/raftstore/store/peer.rs]: you can even use term + index to find the proposal, but here is just my suggest, if we can do, I think it is better to remove uuid now.

Comment (BusyJay) [inline src/raftstore/store/peer.rs]: No, uuid is used to check there is no duplicated proposal has been raised and for client callback validation.

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
CI failed

**Comment by ngaut** [comment]
PTAL @hhkbp2

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**